### ISSUE TITLE
An irrational [E0602] error while using typehint feature in a method member to declare that the return type is the class itself.

### ISSUE THREAD

**Issue by yuriok** [PyCQA/pylint#2552]
### Steps to reproduce
1. Define a class with a method member which will return the current instance(i.e. the 'self');
2. Use the typehint feature to declare that the return type of this method is the Class itself;
3. Then the pyhint will report a [E0602] Undefined variable 'TheClassName' error.

code like this:
```python
class MyClass:
    def return_self(self) -> MyClass:
        return self
```

### Current behavior
The pylint will report a error.

### Expected behavior
The pylint should not report a error

### pylint --version output
2.1.1

**Comment by yuriok** [comment]
I found that this method can solve this problem:
```python
class MyClass:
    def return_self(self) -> self:
        return self
```
But I think it's not a pure solution, because this will mislead the reader while the return value is not the same instance. Like this:
```python
class MyClass:
    def copy(self) -> self:
        result = MyClass()
        # do something
        return result
```
